homeassistant/components/iotawatt/sensor.py

```diff
@@ -0,0 +1,96 @@
+"""Support for IoTaWatt Energy monitor."""
+from __future__ import annotations
+
+from iotawattpy.sensor import Sensor
+
+from homeassistant.components.sensor import SensorEntity, SensorEntityDescription
+from homeassistant.core import callback
+from homeassistant.helpers import entity_registry, update_coordinator
+
+from .const import DOMAIN, ENTITY_DESCRIPTION_KEY_MAP
+from .coordinator import IotawattUpdater
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add sensors for passed config_entry in HA."""
+    coordinator: IotawattUpdater = hass.data[DOMAIN][config_entry.entry_id]
+    created = set()
+
+    @callback
+    def _create_entity(key: str) -> IotaWattSensor:
+        """Create a sensor entity."""
+        created.add(key)
+        return IotaWattSensor(
+            coordinator=coordinator,
+            key=key,
+            mac_address=coordinator.data["sensors"][key].hub_mac_address,
+            name=coordinator.data["sensors"][key].getName(),
+            entity_description=ENTITY_DESCRIPTION_KEY_MAP.get(
+                coordinator.data["sensors"][key].getUnit(),
+                SensorEntityDescription("base_sensor"),
+            ),
+        )
+
+    async_add_entities(_create_entity(key) for key in coordinator.data["sensors"])
+
+    @callback
+    def new_data_received():
+        """Check for new sensors."""
+        entities = [
+            _create_entity(key)
+            for key in coordinator.data["sensors"]
+            if key not in created
+        ]
+        if entities:
+            async_add_entities(entities)
+
+    coordinator.async_add_listener(new_data_received)
+
+
+class IotaWattSensor(update_coordinator.CoordinatorEntity, SensorEntity):
+    """Defines a IoTaWatt Energy Sensor."""
+
+    _attr_force_update = True
+    _attr_icon = "mdi:flash"
+
+    def __init__(self, coordinator, key, mac_address, name, entity_description):
+        """Initialize the sensor."""
+        super().__init__(coordinator=coordinator)
+
+        self._key = key
+        self._attr_unique_id = self._sensor_data.getSensorID()
```

Comment:
The name part of the sensor id, how is that set? What determines the name of the sensor in the library?

Reply:
The is the function from the API library:
```
    def setSensorID(self, hub_mac_address):
        self._sensor_id = hub_mac_address + "_" + self._type + "_" + self._name
```
The name portion of the sensor id could be either the name the user entered in though the device's web interface, such as ```Spa```, or in the case of sensors that have a unit of ```Watts``` the library creates a second sensor for the user appending ```.wh``` to the name. Which in this example the name would be ```Spa.wh``` for the second sensor.

Reply:
yeah that's not a unique ID. That's a user configurable value and cannot be used as an ID. 

I've asked overeasy about unique IDs. For now let's remove it.

Reply:
So I asked overeasy. For input sensors we can use the channel. For output we cannot currently set a unique ID.